File: Java/binary-tree-postorder-traversal.java
//  Binary Tree Postorder Traversal

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(N)
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<Integer>();
        if (root == null)
            return ans;
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode p = root, last = null;
        while (p != null) {
            s.push(p);
            p = p.left;
        }
        while (!s.empty()) {
            p = s.peek();
            if (p.right==null || p.right==last) {
                ans.add(p.val);
                last = p;
                s.pop();
            }
            else {
                p = p.right;
                while (p != null) {
                    s.push(p);
                    p = p.left;
                }
            }
        }
        return ans;
    }
}

File: Leetcode200/binary-tree-postorder-traversal.cc
//  Binary Tree Postorder Traversal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        if (!root)  return res;
        stack<TreeNode *> st;
        TreeNode *p = root, *prev = nullptr;
        while (p) {
            st.push(p);
            p = p->left;
        }
        while (!st.empty()) {
            p = st.top();
            if (!p->right || p->right==prev) {
                st.pop();
                res.push_back(p->val);
                prev = p;
            }
            else {
                p = p->right;
                while (p) {
                    st.push(p);
                    p = p->left;
                }
            }
        }
        return res;
    }
};